floris/parallel_floris_model_2.py

```diff
@@ -0,0 +1,309 @@
+            # the splits, and return them somehow.
+            self._stored_turbine_powers = None # Temporary
+        if self.interface is None:
+            t0 = timerpc()
+            super().run()
+            t1 = timerpc()
+        elif self.interface == "multiprocessing":
+            t0 = timerpc()
+            parallel_run_inputs = self._preprocessing()
+            t1 = timerpc()
+            if self.return_turbine_powers_only:
+                with self._PoolExecutor(self.max_workers) as p:
+                    self._turbine_powers_split = p.starmap(
+                        _parallel_run_powers_only,
+                        parallel_run_inputs
+                    )
+            else:
+                with self._PoolExecutor(self.max_workers) as p:
+                    self._fmodels_split = p.starmap(_parallel_run, parallel_run_inputs)
+            t2 = timerpc()
+            self._postprocessing()
+            self.core.farm.finalize(self.core.grid.unsorted_indices)
+            self.core.state = State.USED
+            t3 = timerpc()
+        if self.print_timings:
+            print("===============================================================================")
+            if self.interface is None:
+                print(f"Total time spent for serial calculation (interface=None): {t1 - t0:.3f} s")
+            else:
+                print(
+                    "Total time spent for parallel calculation "
+                    f"({self.max_workers} workers): {t3-t0:.3f} s"
+                )
+                print(f"  Time spent in parallel preprocessing: {t1-t0:.3f} s")
+                print(f"  Time spent in parallel loop execution: {t2-t1:.3f} s.")
+                print(f"  Time spent in parallel postprocessing: {t3-t2:.3f} s")
+
+    def _preprocessing(self):
+        # Split over the wind conditions
+        n_wind_condition_splits = self.n_wind_condition_splits
+        n_wind_condition_splits = np.min(
+            [n_wind_condition_splits, self.core.flow_field.n_findex]
+        )
+
+        # Prepare the input arguments for parallel execution
+        fmodel_dict = self.core.as_dict()
+        wind_condition_id_splits = np.array_split(
+            np.arange(self.core.flow_field.n_findex),
+            n_wind_condition_splits,
+        )
+        multiargs = []
+        for wc_id_split in wind_condition_id_splits:
+            # for ws_id_split in wind_speed_id_splits:
+            fmodel_dict_split = copy.deepcopy(fmodel_dict)
+            wind_directions = self.core.flow_field.wind_directions[wc_id_split]
+            wind_speeds = self.core.flow_field.wind_speeds[wc_id_split]
+            turbulence_intensities = self.core.flow_field.turbulence_intensities[wc_id_split]
+
+            # Extract and format all control setpoints as a dict that can be unpacked later
+            control_setpoints_subset = {
+                "yaw_angles": self.core.farm.yaw_angles[wc_id_split, :],
```

Comment:
This is bigger than this PR, but something I wonder, will we need a process to go throughout the code should we ever add a new control setpoint?

Reply:
I've been thinking a bit about this too. It would be nice if we could generalize all the control setpoints in some way, so that we don't have to add infrastructure each time we add a new one. They'll likely all be shape n_findex * n_turbines, and we could consider moving to a `control_setpoints` dictionary, which can be unpacked and checked programmatically some way.

Reply:
I do need to add tests for at least some of the control setpoints though, to make sure those are currently handled correctly. Probably just testing with `yaw_angles` is enough for now.

Reply:
I've now added testing for `yaw_angles` (and did need to address a bug!).